### PR TITLE
CentCom Galatic Ban DB

### PR DESCRIPTION
**Pull request by ike709**
Admins will now be able to look up a player's bans from several other servers via the player panel. 

Supported servers: 
* BeeStation 
* /vg/station
* OracleStation
* FTL13
* Fulpstation
* TGMC

Planned support that bobbah is working on:
* Yogstation
* World Server
* Any other server willing to make their bans publicly visible.

API: https://centcom.melonmesa.com
Source: https://github.com/bobbahbrown/CentCom

## Changelog
:cl: ike709 and bobbahbrown
add: Admins can now see your bans on (some) other servers.
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
image

**Comment by MCterra10** [comment]
since when was oracle a server again?

**Comment by ike709** [comment]
@MCterra10 It died in 2018, I asked the host if he still had a copy of the DB and he did.

**Comment by PowerfulBacon** [comment]
This is cool as fuck

**Comment by skull132** [comment]
How long until someone tells leddit :eyes:

**Comment by ike709** [comment]
DNM til I add sanitization.

**Comment by Kmc2000** [comment]
Ascended, SOULful, CHAD html ui

**Comment by TeomanTheGreat** [comment]
"So it says here your wanted for four counts of erp."
This pr is based.

File: config/Sage/config.txt [hunk 1]
```diff
@@ -517,6 +516,10 @@ TOPIC_MAX_SIZE 500
 ## Uncomment to enable Fail2Topic, comment-out to disable
 #TOPIC_ENABLED
 
+## Uncomment to enable global ban DB using the provided URL. The API should expect to receive a ckey at the end of the URL.
+## More API details can be found here: https://centcom.melonmesa.com/swagger/index.html
+#CENTCOM_BAN_DB https://centcom.melonmesa.com/ban/search/
```
Comment (qwertyquerty) [inline config/Sage/config.txt]: Best practice would be to remove the trailing slash here and instead add it in later

File: config/config.txt [hunk 2]
```diff
@@ -519,6 +519,10 @@ TOPIC_MAX_SIZE 500
 ## Enable automatic profiling - Byond 513.1506 and newer only.
 #AUTO_PROFILE
 
+## Uncomment to enable global ban DB using the provided URL. The API should expect to receive a ckey at the end of the URL.
+## More API details can be found here: https://centcom.melonmesa.com/swagger/index.html
+#CENTCOM_BAN_DB https://centcom.melonmesa.com/ban/search/
```
Comment (qwertyquerty) [inline config/config.txt]: Likewise

File: code/modules/admin/topic.dm [hunk 3]
```diff
@@ -1976,6 +1976,62 @@
...
+					dat += "<b>Server: </b> [sanitize(bandata["sourceName"])]<br>"
+					dat += "<b>Type: </b> [sanitize(bandata["type"])]<br>"
+					dat += "<b>Banned By: </b> [sanitize(bandata["bannedBy"])]<br>"
+					dat += "<b>Reason: </b> [sanitize(bandata["reason"])]<br>"
+					dat += "<b>Datetime: </b> [sanitize(bandata["bannedOn"])]<br>"
+					var/expiration = bandata["expires"]
+					dat += "<b>Expires: </b> [expiration ? "[sanitize(expiration)]" : "Permanent"]<br>"
+					if(bandata["type"] == "job")
+						dat += "<b>Jobs: </b> "
+						for(var/job in bandata["jobs"])
```
Comment (qwertyquerty) [inline code/modules/admin/topic.dm]: Just do

```
var/list/jobs = bandata["jobs"]
dat += sanitize(jobs.Join(", "))
```

Fixes trailing comma too

File: code/modules/admin/topic.dm [hunk 4]
```diff
@@ -1976,6 +1976,62 @@
...
+		var/list/bans
+
+		var/dat = "<meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><body>"
+
+		if(response.errored)
+			dat += "<br>Failed to connect to CentCom."
+		else if(response.status_code != 200)
+			dat += "<br>Failed to connect to CentCom. Status code: [response.status_code]"
+		else
+			if(response.body == "[]")
```
Comment (qwertyquerty) [inline code/modules/admin/topic.dm:2007]: Is the DM json parser incapable of handling json lists? If not, this entire section is useless.

Comment (ike709) [reply code/modules/admin/topic.dm:2007]: Seems to be.

Comment (qwertyquerty) [reply code/modules/admin/topic.dm:2007]: I did some tests a bit ago, seemed to work fine for me?

Comment (qwertyquerty) [reply code/modules/admin/topic.dm:2007]: @ike709 oi are you sure

**Comment by ike709** [comment]
Bobbah is going to make a slight API change soon that'll include the RP level of the banning server. DNM this PR until then so I don't have to do a one-line followup PR.

**Comment by BettonnCZ** [comment]
so frequently banned people like me are fucked?

**Comment by ghost** [comment]
inb4 cross-ban because I raided an deathclaw ERP den in DR2

**Review by Crossedfall — APPROVED** [review]

**Comment by ike709** [comment]
Updated picture:
image

**Comment by MCterra10** [comment]
Bobbahbooey

**Comment by bobbah** [comment]
bobbahbooey

**Pull request merged**